azurerm/internal/services/sentinel/sentinel_data_connector_microsoft_defender_advanced_threat_protection.go

```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return err
+	}
+	name := d.Get("name").(string)
+	id := parse.NewDataConnectorID(workspaceId.SubscriptionId, workspaceId.ResourceGroup, workspaceId.WorkspaceName, name)
+
+	if d.IsNewResource() {
+		resp, err := client.Get(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(resp.Response) {
+				return fmt.Errorf("checking for existing Sentinel Data Connector Microsoft Defender Advanced Threat Protection %q: %+v", id, err)
+			}
+		}
+
+		id := dataConnectorID(resp.Value)
+		if id != nil && *id != "" {
+			return tf.ImportAsExistsError("azurerm_sentinel_data_connector_microsoft_defender_advanced_threat_protection", *id)
+		}
+	}
+
+	tenantId := d.Get("tenant_id").(string)
+	if tenantId == "" {
+		tenantId = meta.(*clients.Client).Account.TenantId
+	}
+
+	param := securityinsight.MDATPDataConnector{
+		Name: &name,
+		MDATPDataConnectorProperties: &securityinsight.MDATPDataConnectorProperties{
+			TenantID: &tenantId,
+			DataTypes: &securityinsight.AlertsDataTypeOfDataConnector{
+				Alerts: &securityinsight.AlertsDataTypeOfDataConnectorAlerts{
+					State: securityinsight.Enabled,
+				},
+			},
+		},
+		Kind: securityinsight.KindMicrosoftDefenderAdvancedThreatProtection,
+	}
+
+	_, err = client.CreateOrUpdate(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, id.Name, param)
+	if err != nil {
+		return fmt.Errorf("creating Sentinel Data Connector Microsoft Defender Advanced Threat Protection %q: %+v", id, err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceSentinelDataConnectorMicrosoftDefenderAdvancedThreatProtectionRead(d, meta)
+}
+
+func resourceSentinelDataConnectorMicrosoftDefenderAdvancedThreatProtectionRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Sentinel.DataConnectorsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.DataConnectorID(d.Id())
+	if err != nil {
+		return err
+	}
+	workspaceId := loganalyticsParse.NewLogAnalyticsWorkspaceID(id.SubscriptionId, id.ResourceGroup, id.WorkspaceName)
+
+	resp, err := client.Get(ctx, id.ResourceGroup, operationalInsightsResourceProvider, id.WorkspaceName, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] Sentinel Data Connector Microsoft Defender Advanced Threat Protection %q was not found - removing from state!", id)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("retrieving Sentinel Data Connector Microsoft Defender Advanced Threat Protection %q: %+v", id, err)
+	}
+
+	if err := assertDataConnectorKind(resp.Value, securityinsight.DataConnectorKindMicrosoftDefenderAdvancedThreatProtection); err != nil {
+		return fmt.Errorf("asserting Sentinel Data Connector Microsoft Defender Advanced Threat Protection of %q: %+v", id, err)
+	}
+	dc := resp.Value.(securityinsight.MDATPDataConnector)
```

Comment:
we can remove the `assert` function here and reduce this to:

```suggestion
	dc, ok := resp.Value.(securityinsight.MDATPDataConnector)
	if !ok {
		return fmt.Errorf("%s was not a Microsoft Defender Advanced Threat Protection", *id, err)
	}
```

Reply:
The assert function can be used both in Read function and the `importer`. For the sake of not to repeat the type assertion among all the DC files, I'd rather centralize the logic in one place.